<?xml version="1.0"?>
<robot name="car_model" xmlns:xacro="http://www.ros.org/wiki/xacro">

        <material name="black">
                <color rgba="0 0 0 1"/>
        </material>
        <material name="blue">
                <color rgba="0 0 0.8 1"/>
        </material>
        <material name="white">
                <color rgba="1 1 1 1"/>
        </material>
        <material name="gray">
                <color rgba="0.3 0.3 0.3 1"/>
        </material>

        <xacro:property name="wheelradius" value="0.125" />
        <xacro:property name="baselength" value="0.5" />
        <xacro:property name="baseheight" value="${wheelradius}" />
        <xacro:property name="axlelength" value="0.4" />
        <xacro:property name="update_rate" value="1.0" />

        <xacro:property name="base_mass" value="1e-1" />
        <xacro:property name="axle_mass" value="${0.5*1e-1}" />
        <xacro:property name="wheel_mass" value="1e+0" />
        <xacro:property name="camera_mass" value="1e-2" />

        <xacro:macro name="default_inertial" params="mass">
                <inertial>
                        <mass value="${mass}" />
                        <inertia ixx="${mass*0.01}" ixy="0.0" ixz="0.0" iyy="${mass*0.01}" iyz="0.0" izz="${mass*0.01}" />
                </inertial>
        </xacro:macro>

        <xacro:macro name="world_link">
                <link name="base_footprint">
                        <visual>
                                <geometry>
                                        <box size="0.001 0.001 0.001"/>
                                </geometry>
                                <origin rpy="0 0 0" xyz="0.001 0.001 0.001"/>
                        </visual>
                        <collision>
                                <geometry>
                                        <box size="0.001 0.001 0.001"/>
                                </geometry>
                        </collision>
                        <xacro:default_inertial mass="${base_mass/100}"/>
                </link>
        </xacro:macro>


        <xacro:macro name="car_base">
                <link name="base_link">
                        <visual>
                                <geometry>
                                        <box size="${baselength/2} ${baselength} ${baselength/8}" />
                                </geometry>
                                <material name="black" />
                        </visual>
                        <collision>
                                <geometry>
                                        <box size="${baselength/2} ${baselength} ${baselength/8}" />
                                </geometry>
                        </collision>
                        <xacro:default_inertial mass="${base_mass}"/>
                </link>
                <joint name="base_footprint_joint" type="fixed">
                        <axis xyz="0 0 1" />
                        <parent link="base_footprint" />
                        <child link="base_link" />
                        <origin rpy="0 0 0" xyz="0 0 ${baseheight}" />
                </joint>
        </xacro:macro>


        <xacro:macro name="car_axle" params="frontback position">
               <link name="${frontback}_axle">
                       <visual>
                               <geometry>
                                       <cylinder radius="${baselength/8}" length="${axlelength}" />
                               </geometry>
                               <material name="gray" />
                       </visual>
                       <collision>
                               <geometry>
                                       <cylinder radius="${baselength/8}" length="${axlelength}" />
                               </geometry>
                       </collision>
                       <xacro:default_inertial mass="${axle_mass}"/>
               </link>
               <joint name="${frontback}_axle_to_base" type="fixed">
                       <axis xyz="1 0 0" />
                       <mimic joint="base_footprint_joint" multiplier="1"/>
                       <parent link="base_footprint" />
                       <child link="${frontback}_axle" />
                       <origin rpy="0 -1.57 0" xyz="0 ${position * 0.3} ${baseheight}" />
               </joint>
       </xacro:macro>


        <xacro:macro name="car_front_wheel" params="leftright lr axle position upper lower">
                <link name="front_${leftright}_wheel">
                        <visual>
                                <geometry>
                                        <cylinder radius="${wheelradius}" length="${wheelradius}" />
                                </geometry>
                                <material name="white" />
                        </visual>
                        <collision>
                                <geometry>
                                        <cylinder radius="${wheelradius}" length="${wheelradius}" />
                                </geometry>
                        </collision>
                        <xacro:default_inertial mass="${wheel_mass}"/>
                </link>
                <joint name="front_${leftright}_wheel_to_front_axle" type="revolute">
                        <axis xyz="1 0 0" />
                        <limit lower="${lower}" upper="${upper}" effort="1000.0" velocity="1.0"/>
                        <parent link="base_footprint" />
                        <child link="front_${leftright}_wheel" />
                        <origin rpy="0 -1.57 0" xyz="${-lr} 0.3 ${baseheight}" />
                </joint>
        </xacro:macro>


        <xacro:macro name="car_back_wheel" params="leftright axle lr upper lower">
                <link name="back_${leftright}_wheel">
                        <visual>
                                <geometry>
                                        <cylinder radius="${wheelradius}" length="${wheelradius}" />
                                </geometry>
                                <material name="white" />
                        </visual>
                        <collision>
                                <geometry>
                                        <cylinder radius="${wheelradius}" length="${wheelradius}" />
                                </geometry>
                        </collision>
                        <xacro:default_inertial mass="${wheel_mass}"/>
                </link>

                <joint name="back_${leftright}_wheel_to_back_axle" type="continuous">
                        <axis xyz="0 0 1" />
                        <parent link="base_footprint" />
                        <child link="back_${leftright}_wheel" />
                        <origin rpy="0 -1.57 0" xyz="${-lr} -0.3 ${baseheight}" />
                </joint>
        </xacro:macro>


        <xacro:macro name="camera_link" params="height width">
                <link name="camera">
                        <visual>
                                <geometry>
                                        <box size="0.1 0.1 0.1"/>
                                </geometry>
                                <material name="black"/>
                        </visual>
                        <collision>
                                <geometry>
                                        <box size="0.1 0.1 0.1"/>
                                </geometry>
                        </collision>
                        <xacro:default_inertial mass="${camera_mass}"/>
                </link>
                <joint name="camera_joint" type="fixed">
                        <axis xyz="0 0 1" />
                        <origin xyz="0 ${(baselength/2)+(baselength/4)} ${baseheight}" rpy="0 0 1.57"/>
                        <parent link="base_footprint"/>
                        <child link="camera"/>
                </joint>
        </xacro:macro>


        <xacro:world_link />
        <xacro:car_base />
        <xacro:camera_link height="240" width="320" />
        <xacro:car_axle frontback="front" position="1" />
        <xacro:car_axle frontback="back" position="-1" />

        <xacro:car_front_wheel leftright="left" lr="${(axlelength/2) + 0.05}" axle="1"
                position="1" upper="0.785" lower="-0.393" />
        <xacro:car_front_wheel leftright="right" lr="${-(axlelength/2) - 0.05}" axle="1"
                position="-1" upper="0.393" lower="-0.785" />

        <xacro:car_back_wheel leftright="left" lr="${(axlelength/2) + 0.05}" axle="-1"
                upper="${3.14*1000}" lower="${-3.14*1000}" />
        <xacro:car_back_wheel leftright="right" lr="${-(axlelength/2) - 0.05}" axle="-1"
                upper="${3.14*1000}" lower="${-3.14*1000}" />

</robot>
